File: 11vaFeriaAcademicayCultural/Intermedio/Problema4/problema4.cpp
/*Problema 4 Without Repetitions
Propuesta de solucion
autor: Luis Arturo De La Garza Navel
C++*/
//Nota: esta solucion no elimina las repeticiones del caracter "¡"
#include <iostream>
#include<string>
#include<vector>
#include<locale>
using namespace std;

int main() {
  string oracion;
  getline(cin, oracion);
  vector<char> nueva_oracion;//vector para la nueva oracion
  bool mayuscula = isupper(oracion[0]); //bandera para verificar si hemos insertado una letra mayuscula
  char caracter_actual = oracion[0], caracter; //variable para comparar las repeticiones contiguas
  nueva_oracion.push_back(caracter_actual);
  for(int i = 1; i < oracion.size(); i++){ //iniciamos del 2do caracter
    caracter = tolower(oracion[i]);
    if(caracter != caracter_actual){
      if(isupper(oracion[i]) && not mayuscula){
        caracter = oracion[i];//cambiamos el caracter a mayuscula para insertarlo en la lista
        mayuscula = true;
      }
      cout << caracter_actual << '\n';
      nueva_oracion.push_back(caracter);
      caracter_actual = caracter;
    }
  }
  for(int j = 0; j < nueva_oracion.size(); j++){cout <<nueva_oracion[j];}
  cout <<endl;
  return 0;
}

File: 11vaFeriaAcademicayCultural/Intermedio/Problema4/problema4.py
"""Problema 4 Without Repetitions
Propuesta de solucion
autor: Luis Arturo De La Garza Navel
Python 2.7"""

"""el uso del modulo sys para entrada/salida de datos es altamente recomendado en programacion
competitiva (ojo: no todas las competencias permiten su uso), pero puede usar los metodos
propios definidos en python (raw_input y print)"""
from sys import stdin, stdout

if __name__ == '__main__':
    oracion = stdin.readline().decode("utf8") #  decodificamos los carcateres de utf8 para manipularnos
    nueva_oracion = [] #  lista para la nueva oracion
    mayuscula = oracion[0].isupper() # bandera para verificar si hemos insertado una letra mayuscula
    caracter_actual = oracion[0] #  variable para comparar las repeticiones contiguas
    nueva_oracion.append(caracter_actual)
    for _caracter in oracion[1:]: #  iniciamos del 2do caracter
        caracter = _caracter.lower() #  convertimos el caracter a minuscula para comparar
        if caracter != caracter_actual:
            if _caracter.isupper() and not mayuscula:
                caracter = caracter.upper() #  cambiamos el caracter a mayuscula para insertarlo en la lista
                mayuscula = True
            nueva_oracion.append(caracter)
            caracter_actual = caracter
    #  codificamos los carcateres de utf8 para imprimir la cadena final
    stdout.write('{}'.format(''.join(nueva_oracion).encode("utf8")))
